### DISCUSSION TITLE
Use yarn 3.x pnp with Remix framework

### DISCUSSION THREAD

**Discussion by meglio** [yarnpkg/berry#4874] [Q&A] [answered]
Remix version:  1.7.1

### Steps to Reproduce

Create a simple test.jsx Remix route:

```jsx
import { json } from "@remix-run/node";

export default function SomePage(props) {
    return <div>
        Page
    </div>
}
```

Then `yarn dev`.

### Expected Behavior

No errors in developer console.

### Actual Behavior

Error in developer console:

![image](https://user-images.githubusercontent.com/282177/191274209-88932318-4bf0-4f9d-a9d3-b3ae7365e3d7.png)

### package.json

```json
{
  "private": true,
  "sideEffects": false,
  "scripts": {
    "build": "remix build",
    "dev": "remix dev",
    "start": "remix-serve build"
  },
  "dependencies": {
    "@remix-run/node": "^1.7.1",
    "@remix-run/react": "^1.7.1",
    "@remix-run/serve": "^1.7.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "devDependencies": {
    "@remix-run/dev": "^1.7.1",
    "@remix-run/eslint-config": "^1.7.1",
    "eslint": "^8.20.0"
  },
  "engines": {
    "node": ">=14"
  },
  "packageManager": "yarn@3.2.3"
}


```

### Yarn and Node

* node version: 18.9.0
* Yarn: 3.x

NB. I tried creating the project with using Yarn 1.x and it worked fine. The problem appears after migrating to yarn 3.x

Might be related:

https://github.com/yarnpkg/berry/issues/638#issuecomment-1124629341

Looks like Yarn 3.x supports ESM module loading.  So I tried the following in my `.yarnrc.yml` file, following the recommendations from the link above:

```yml
yarnPath: .yarn/releases/yarn-3.2.3.cjs
pnpEnableEsmLoader: true
nodeLinker: "pnp"
```

But this did not help.

So, the thread above makes me think that yarn 3.x should work with Remix. But it does not. I'm wondering if this is Remix or Yarn issue?

**Comment by RDIL** [comment]
I tried to reproduce (https://github.com/RDIL/remix-pnp-test) but was unable to.  Can you try with my repository and see if it works for you?  If it does, it sounds like something isn't right with your specific setup.

**Comment by meglio** [comment]
@RDIL looking at your code, can't find where you do
`import { json } from "@remix-run/node";`

Could you please double-check you have a `test.jsx` like the one from my original report, and then try accessing that page in your browser. That's when the error is thrown.

Just create that page in the `routes` folder.

**Reply by RDIL** [reply]
Huh, yeah, that does cause it.

I don't think this is PnP's fault though.  It looks like remix is doing a lot of really strange stuff at runtime to get this all working.  I think it's worth opening an issue in the Remix repository, as I don't entirely understand how it's doing things/where it is even trying to access `fs/promises` from.

**Comment by meglio** [comment]
Let's try: https://github.com/remix-run/remix/issues/4236

**Comment by meglio** [comment, marked as answer]
@RDIL , there's an update in that issue, could you please check if it makes any difference? Thank you.

https://github.com/remix-run/remix/issues/4236#issuecomment-1264444539

**Reply by RDIL** [reply]
Yeah, that seems to fix it.  See https://github.com/RDIL/remix-pnp-test/commit/9be5fafd9072ecbd111c7b17460842113869c102 for a patch that can resolve the issue.

**Reply by meglio** [reply]
Great, so is this something the Yarn team would take over and complete, or is it something that the Remix team needs to improve in their code?

**Reply by RDIL** [reply]
It's on Remix's end, not ours.
